api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,570 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package websocketupgradeproto
+
+import (
+	"context"
+	"io"
+	"log/slog"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gobwas/ws"
+
+	"github.com/gravitational/teleport/api/constants"
+)
+
+const (
+	// ComponentTeleport is the name of the Teleport server component.
+	ComponentTeleport = "teleport"
+	// ComponentClient is the name of the Teleport client component.
+	ComponentClient = "client"
+)
+
+// connectionType represents the side of the connection.
+// It can be either a server or a client connection.
+// This is used to determine how to handle WebSocket frames, especially
+// for the ping/pong frames.
+type connectionType int
+
+const (
+	// serverConnection indicates that this is a server-side connection.
+	serverConnection connectionType = iota + 1
+	// clientConnection indicates that this is a client-side connection.
+	clientConnection
+)
+
+const (
+	// maxDeadlineOnClose is the time to wait for the close handshake to complete
+	// before forcefully closing the underlying connection.
+	maxDeadlineOnClose = 3 * time.Second
+)
+
+var _ interface {
+	net.Conn
+	NetConn() net.Conn
+} = (*Conn)(nil)
+
+// Conn represents a WebSocket connection that implements the net.Conn interface
+// backed by an underlying WebSocket connection to bypass proxies and load balancers
+// that terminate TLS connections between the client and server.
+// It provides methods for reading and writing WebSocket frames, managing control
+// frames (Ping, Pong, and Close), and handling connection state.
+//
+// The Conn type ensures compliance with the WebSocket close handshake, allowing
+// for graceful termination of the connection while maintaining protocol integrity
+// and proper resource cleanup.
+type Conn struct {
+	managedConn
+
+	// underlyingConn is the underlying network after the WebSocket upgrade.
+	underlyingConn net.Conn
+	// logContext is the context used for logging.
+	logContext context.Context
+	// logger is the logger used for logging.
+	logger *slog.Logger
+	// connType indicates whether this is a server or client connection.
+	// This affects how ping/pong frames are handled as clients never send pings
+	// and servers never respond to pongs.
+	connType connectionType
+	// protocol is the negotiated WebSocket sub-protocol for this connection.
+	protocol string
+	// supportsCloseProccess indicates whether this connection supports
+	// the WebSocket close process.
+	supportsCloseProccess bool
+
+	// pingPongReplies holds the ping/pong frames to be sent in the next write loop iteration.
+	pingPongReplies []ws.Frame
```

Comment:
As per https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets we can (should?) only keep track of the last pong to be sent.